### PR TITLE
ShaderMask needs to play nicely with compositing

### LINKED ISSUE #1155: ShaderMask needs to play nicely with compositing
Something like the following code used to work, and now no longer works.

This code should put a fading edge at the top of the block so the children fade away to the background color as they go under the mask.  Now the entire block goes away.  In my use case I see a chunk of area that is fully faded away that isn't faded away when just the shader mask is removed.

```dart
Future main() async =>
  runApp(new Container(
      decoration: new BoxDecoration(backgroundColor: new Color(0xFF808080)),
      child: new ShaderMask(
          shaderCallback: _createShader,
          child: new Block([
            _createColoredBox(0xFFFF0000),
            _createColoredBox(0xFF0FF000),
            _createColoredBox(0xFF0000FF),
            _createColoredBox(0xFFFFFF00)
          ]))));

Widget _createColoredBox(int hexColor) => new Container(
    height: 300.0,
    decoration: new BoxDecoration(backgroundColor: new Color(hexColor)));

// Fading edge colors need to be pure white with some value of alpha to properly fade the children.
ui.Shader _createShader(Rect bounds) => new LinearGradient(
    begin: Point.origin,
    end: new Point(0.0, 100.0),
    colors: [const Color(0x00FFFFFF), const Color(0xFFFFFFFF)],
    stops: [0.0, 1.0]).createShader();
```

### LINKED ISSUE #1177: Implement RenderStatisticsBox.alwaysNeedsCompositing
(derived from a review comment by @abarth on https://github.com/flutter/flutter/pull/1166)

Currently `RenderStatisticsBox` lies because it creates a layer but doesn't have the compositing bits set to say that it creates a layer. We should either set `hasLayer` to true, or explicitly set it to false with a comment explaining why.

The risk is that turning on the performance overlay can change the compositing strategy for the app. If you leave it off, but put the overlay deep in your tree, you can end up breaking things like opacity that are higher up the tree.

In practice you almost always put it at the top of the tree.

### PR DESCRIPTION
**Pull request by abarth**
This patch adds ShaderLayer and makes ShaderMask use it.

Fixes #1155

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @chinmaygarde 

Needs engine roll before landing.

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -298,27 +298,19 @@ class PaintingContext {
...
-     ..transferMode = transferMode
-     ..shader = shaderCallback(bounds);
-  }
-
   /// Push a shader mask.
   ///
   /// This function will call painter synchronously with a painting context that
   /// will be masked with the given shader.
   ///
   /// WARNING: This function does not yet support compositing.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: the comment lies!

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -540,7 +540,7 @@ class RenderOpacity extends RenderProxyBox {
     assert(opacity >= 0.0 && opacity <= 1.0);
   }
 
-  bool get alwaysNeedsCompositing => _alpha != 0 && _alpha != 255;
+  bool get alwaysNeedsCompositing => child != null && (_alpha != 0 && _alpha != 255);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: you need to update the bits when child changes from null to non-null or vice-versa

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 3]
```diff
@@ -604,9 +604,14 @@ class RenderShaderMask extends RenderProxyBox {
     markNeedsPaint();
   }
 
+  bool get alwaysNeedsCompositing => child != null;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: you need to update the bits when child changes from null to non-null or vice-versa

**Comment by Hixie** [comment]
lgtm

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 3 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: RenderObject itself does that for us.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: ah right.

**Pull request merged**